packages/web/nexus/subscription.ts

```diff
@@ -0,0 +1,322 @@
+    // Payment failed
+    throw new Error("Unable to update subscription, possible payment failure")
+  }
+  await stripe.subscriptions.update(stripeSubscription.id, {
+    cancel_at_period_end: cancelAtPeriodEnd,
+  })
+  // Update our records with the new subscription info
+  return db.membershipSubscription.update({
+    where: {
+      userId,
+    },
+    data: {
+      period: subscriptionPeriod,
+      expiresAt: new Date(subscriptionUpdated.current_period_end * 1000 + (24 * 60 * 60 * 1000 * 2)),
+      stripeSubscription: subscriptionUpdated as unknown as InputJsonValue,
+    },
+  })
+}
+
+const MembershipSubscriptionMutations = extendType({
+  type: 'Mutation',
+  definition(t) {
+    t.field('purchaseMembershipSubscription', {
+      type: 'MembershipSubscription',
+      args: {
+        period: arg({ type: 'MembershipSubscriptionPeriod', required: true }),
+        token: stringArg({ required: true }),
+      },
+      resolve: async (_parent, args, ctx) => {
+        const { userId } = ctx.request
+
+        if (!userId) {
+          throw new Error("You must be logged in to create a subscription")
+        }
+
+        const user = await ctx.db.user.findUnique({
+          where: {
+            id: userId,
+          },
+          include: {
+            membershipSubscription: true,
+          }
+        })
+
+        if (!user) throw new Error("User not found")
+
+        let customerId
+        if (!user.stripeCustomerId) {
+          const customer = await stripe.customers.create({
+            payment_method: args.token,
+            description: `${user.handle} (${user.id})`,
+            email: user.email,
+            metadata: {
+              id: user.id,
+              handle: user.handle,
+            },
+          })
+          customerId = customer.id
+          await ctx.db.user.update({
+            where: {
+              id: userId,
+            },
+            data: {
+              stripeCustomerId: customerId,
+            },
+          })
+
+          const stripeSubscription = await stripe.subscriptions.create({
+            items: [{
+              price: getSubscriptionPriceId(args.period),
+              metadata: {
+                journalyUserId: userId,
+              },
+            }],
+            default_payment_method: args.token,
+            customer: customerId,
+          })
+  
+          const subData = {
+            period: args.period,
+            // Give 2 days grace period
+            expiresAt: new Date(stripeSubscription.current_period_end * 1000 + (24 * 60 * 60 * 1000 * 2)),
+            // We weren't smart enough to make TS know that Stripe.Response & InputJsonValue are comparable :'(
+            stripeSubscription: stripeSubscription as unknown as InputJsonValue,
+            stripeSubscriptionId: stripeSubscription.id,
+            lastFourCardNumbers: '1234',
+            cardType: 'Visa',
+          }
+  
+          // TODO: Log failure and get proper alarms set up
```

Comment:
heh, we should talk about doing this. I'd like to have this set up by the time we ship.